src/framework/core/js/ModelTransformationTransforms.js

```diff
@@ -349,86 +349,188 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
             newOptions[path] = oldKey;
         });
-        transformSpec.options = newOptions;
+        if (!$.isEmptyObject(newOptions)) {
+            transformSpec.options = newOptions;
+        }
         return transformSpec;
     };
 
-    fluid.transforms.arrayToSetMembership.invert = function (transformSpec, transformer) {
-        return fluid.transforms.arrayToSetMembership.invertWithType(transformSpec, transformer,
-            "fluid.transforms.setMembershipToArray");
-    };
-
     fluid.defaults("fluid.transforms.setMembershipToArray", {
         gradeNames: ["fluid.standardTransformFunction", "fluid.lens"],
         invertConfiguration: "fluid.transforms.setMembershipToArray.invert"
     });
 
-    fluid.transforms.setMembershipToArray = function (input, transformSpec, transformer) {
-        var options = transformSpec.options;
-
-        if (!options) {
-            fluid.fail("setMembershipToArray requires an options block specified");
+    /**
+     *
+     * Transform the input object into an array based upon the options provided to the specification. Accepts
+     * an object as its first argument and an optional second argument (an object) which contains the
+     * specifications for the resulting output.
+     *
+     * An error will be thrown if the input to be transformed is not an object.
+     *
+     * Partly invertible (into arrayToSetMembership)
+     *
+     * For example, with the following transformSpec:
+     *
+     * {
+     *   "options": {
+     *      "hasMouse": "mouse",
+     *      "hasKeyboard": "keyboard",
+     *      "hasTrackpad": "trackpad",
+     *      "hasHeadtracker": "headtracker"
+     *   },
+     *   "presentValue": "supported",
+     *   "missingValue": "not supported"
+     * }
+     *
+     * For an input object like this:
+     *
+     * {
+     *   "hasMouse": "supported",
+     *   "hasKeyboard": "supported",
+     *   "hasTrackpad": "not supported",
+     *   "hasHeadtracker": "not supported"
+     * }
+     *
+     * The output array will be:
+     *
+     * ["keyboard", "mouse"]
+     *
+     * https://docs.fluidproject.org/infusion/development/ModelTransformationAPI.html#fluidtransformssetmembershiptoarray
+     *
+     * @param {Object} input - The object to be transformed into an array.
+     * @param {Object} [transformSpec] - (optional) The options provided to the transformation rule.
+     * @return {Array} - The transformed array
+     *
+     */
+    fluid.transforms.setMembershipToArray = function (input, transformSpec) {
+        // <input> should be an object.
+        if (!fluid.isPlainObject(input, true)) {
+            fluid.fail("setMembershipToArray didn't find object at inputPath nor passed as value.");
         }
-
-        if (transformSpec.presentValue === undefined) {
-            transformSpec.presentValue = true;
+        var outputArr = [];
+        var presentValue = fluid.get(transformSpec, "presentValue") ? transformSpec.presentValue : true,
```

Comment:
This code is certainly an improvement over the previous implementation which modified the caller's arguments but my personal preference is for a transformSpec = transformSpec || {} at the start and then returning all of these to being transformSpec.options, and presentValue = transformSpec.presentValue || true etc. which are considerably more readable and performant. fluid.get is an extremely heavyweight piece of machinery and not suitable for use in low-level framework code (in practice, when the framework really needs something like the full functionality it uses various cheaper, semi-private alternatives such as fluid.getImmediate, fluid.model.getSimple etc.)

Reply:
Hi @amb26, thanks for the feedback. I'll remove the usage of `fluid.get` and use the normal JavaScript pattern. I have one doubt though

> my personal preference is for a transformSpec = transformSpec || {} at the start and then returning all of these to being transformSpec.options, and presentValue = transformSpec.presentValue || true etc. which are considerably more readable and performant.

Wouldn't it be more effective if we directly access it from the argument instead of `transformSpec = transformSpec || {}` at the beginning?

Reply:
What do you mean by access it directly from the argument, given that the argument might be undefined?

Reply:
I don't think that the `transformSpec` will be `undefined` in cases when it is used for model transformation (not sure if it is used somewhere else too). The `transformSpec` will always have at least one key namely `type`. So `transformSpec` shouldn't be `undefined` and we can use it safely.

Coming to our case, if we try to access it by `transformSpec.presentValue` and `transformSpec.missingValue`, it would simply return `undefined` in cases where it wasn't provided to the transformation rule. But `transformSpec` will never be undefined (in case of model transformations).

To verify this, I set up a model component (with subcomponents) for both the transformations and passed nothing else other than the transformation `type` to the transformation rule. I tried logging the value to the console from inside the transformation functions. It showed the following results:
![transformSpec](https://user-images.githubusercontent.com/37256767/83274965-1f6fdd00-a1ec-11ea-8dd1-90f3f0ef8c9f.png)

Reply:
One of the use cases for this pull was to allow the transformSpec to be omitted so that the function is useful as a standalone utility - and you have documented it as optional in your JSDoc!

Reply:
Sorry, I almost forgot about that use case. I now understand why we should use `transformSpec = transformSpec || {}` here Thank you for clarifying.